Update shared dtx

Paul-Baudin opened the pull request:
Hi Stephen,

Here are some questions and some little modifications I would like to do, to better suit to my project. Let me know what do you think is good/bad :)

- It is not clear for me what would be the best way of dealing with artefacts :
Now in all the EpiCode structures we will have an 'artefact' field in trialinfo. I would think than the better for me would be simply to completely reject artefacted trials, using ft_selectdata and ft_spike_select after calling the EpiCode reading functions. And then, if we want to know which trial corresponds to which (for example between LFP and spikes), we can use the 'trialnr' in trialinfo which will keep the initial index. I already had to do it previously when comparing LFP trials and spike trials, because in readLFP.m I have some trials which do not load because they are cut into 2 dirs (so it is removed by `full_trial = Startsample > 0 & Endsample < length(dat.trial{1});`), whereas for spikes, because data is concatenated, the same trials load.
We already discussed about it, when rejecting spike trials I prefere to completely reject it, ie the trials are re-indexed in spikedata.trial, and trials are removed from trialinfo (othewise trial is still here and just 'emptied', so considered as a trial with no spike, which I do not want). For that, I use my version of ft_spike_select, which I renamed ft_spike_select_rmalltrials in the /development/modified_fieldtrip_functions folder. I still have the initial index of the trial in trialinfo.trialnr to have correspondance between several data structures.

- readLFP : For consistency between structures, I added the artefact detection in readLFP, let me know if it is fine for you.

- I made small modifications in the artefact rejection code (common for readSpikeTrails_MuseMarkers and read SpikeTrials_windowed, and now readLFP) :
- [ ] For finding artefacts, in readSpikeTrials_MuseMarkers, I had a bug because the variable hypend did't exist for me. I suggest an other way of finding artefacts, I think it could be more readable : if the trial is entirely before, or entirely after the artefacts, then (and only then) it is not artefacted. I applied it in the 3 functions.
- [ ] To avoid too big output  while searching artefacts overlaps, I replaced fprintf by ft_progress.
- [ ] Also, for Spyking-Circus analysis, I removed lot of very short artefacts, so I do not want sometimes to count them as artefacts in the analysis (the rejected period can be considered as negligible). So I added an optional cfg.spike.minbadtime.(markername), to ignore artefact if it is too short.
- [ ] To avoid to have this code copied 3 times, would it be better to have a function findartefactsMuse, which loops over parts and markernames and does that ? And with an optional input to specify if we want to keep artefacted trials or to remove it with ft_selectdata if it is LFP data, or ft_spike_select or ft_spike_select_rmalltrials if it is spike data ?

- Finding header in readSpikeTrials_MuseMarkers and readSpikeTrials_waveform : I would think it is simpler to directly use the one stored in SpikeRaw and not having to again search for datafile and ft_read_header ?

- readSpikeTrials_windowed : it is nice for me and cleaner that what I did. But to be able to use it, I had to add a cfg option to choose whether I want to find overlap with hypnogram or IED. I set the default as 'yes' so it those not break your script.

- readMuseMarkers : I was annoyed by too much output, so I replaced the fprintf by ft_info, so I can control if I want to have it or not.

- readSpikeRaw_Phy :
- [ ] **Attention I had a bug than could have consquences for your analysis.** While looping over channames, if you do not clean SpikeRaw and have less units in the following channame, the last units are remaining from the last channame. I added a `clear SpikeRaw` at the end of the loop to avoid that.
- [ ] to load data before having read it on Phy, the phydata.cluster_group must be read only if 'ischecked' so I made this modification.
- [ ] I had a bug with `chandir = string(channelname{2:end})`, it didn't work, I had to replace it by `chandir = string(channelname(2:end))` with () instead of {}. Did you have this issue too ?

- spikeTrialStats : I added an option to ignore hypnogram, again with a default that does not break your code. I added config input to define ISI histogram bins, and I added the auto-correlogram for each unit.

- writeSpykingCircusDeadfilesMultichannel : I added an optional suffix, because I want 2 dead files : one with the seizures as BAD, and an other without. Then when I launch Spyking Circus I select which one I want to use.

Paul-Baudin committed:
Merge branch 'update_dtx_project_nov'

Paul-Baudin committed:
Merge branch 'master' of https://github.com/stephenwhitmarsh/EpiCode

Paul-Baudin committed:
update shared functions for dtx project

Paul-Baudin committed:
Merge branch 'master' of https://github.com/stephenwhitmarsh/EpiCode into update_shared_dtx

Paul-Baudin committed:
Merge branch 'master' of https://github.com/stephenwhitmarsh/EpiCode into update_shared_dtx

Paul-Baudin committed:
trialinfo starttime and endtime are the clock time of trial begin-end and not time of the marker

Paul-Baudin commented:
I just updated stattime and endtime of timelocked data in trialinfo (readLFP and readSpikeTrials_MuseMarkers) : previously it was just the clock time of the marker, now it is the clock time of the begining and the end of the trial.

Paul-Baudin committed:
add 'sample' field for trial definition and selection, and fix bug with artefact detection

Paul-Baudin committed:
fix bugs in readSpikeWaveforms

Paul-Baudin committed:
find if hyplabel field exists, and if hdr field exists

Paul-Baudin committed:
test a new way of finding artefact overlaps

Paul-Baudin committed:
delete dupplicates created while committed with change names

Paul-Baudin committed:
add external binary search

Paul-Baudin committed:
Merge branch 'update_shared_dtx' of https://github.com/Paul-Baudin/EpiCode into update_shared_dtx

Paul-Baudin commented:
As we discussed I did the following modifications :
- find if hdr field exists in input structure, and if so, do not read it from data file.
- search if hyplabel field exists to know wheter to do hypnogram analysis or not

Also, I did other modifications :
- add the 'sample' field in the modified fieldtrip function ft_spike_maketrials
- correct bugs in readSpikeWaveforms to find the good channel to load, if channelname is not empty
- I tried a new way of searching for artefacts, with binary search. I tried it in readSpikeTrials_MuseMarkers. It takes < 1s instead of several minutes, and I find exactly the same artefacted trials. For now the function is unchanged, except for 2 more 'test' fields in trialinfo. Can you try it after merging this PR and tell me if it works for you ? If so I will replace the artefact detections in the 3 read functions.
- Also, only in readSpikeTrials_MuseMarkers for now, I added a field with the artefact length, so I can then ignore artefacts if they are too short, afterwards.

Cheers

stephenwhitmarsh reviewed (approved):
Okay, I didn't know this function. Indeed, to make it consistent (kinda) I changed, when needed, .samples to .sample. In principle, .samples should not be neccecary anymore. I will use Atom to check if it gets used in other places, and if not, can I remove this?

Paul-Baudin commented:
Yes :)

stephenwhitmarsh commented on shared/readSpikeTrials_MuseMarkers.m:
```diff
@@ -241,9 +243,72 @@
+        
+        %%
+        
+        cfg.spike.minbadtime.(markername) = 0;%ft_getopt(cfg.spike.minbadtime, char(markername), 0);
```
this should be uncommended, or rather removed, because minbadtime is seomthing to take care of later, outside function, right?

stephenwhitmarsh reviewed (commented):
Cool if this speeds things up. However, since this is based on a MEX file, if I understand it correctly, which means its compiled for certain platforms, but might easily become outdated or incompatible, using this function should be done with case, and in the case it doesn't work, functions that use it should be able to revert back to a native MATLAB function.
So there should be 
A) a check for the external, addpath if present, revert to other option if not (ideally with warning).
B) the previous step can be done by switching a default setting, that can also be set in the cfg (but defaults to the mex file).
*** NOW that I look at it again, it might be that it already has the m-file as a 'backup'. In that case, it should be fine!

Paul-Baudin committed:
change binary serch external from mex to matlab

Paul-Baudin committed:
remove minbadtime and update artefact detection

Paul-Baudin commented:
Hi Stephen,
You are right, the mex file was not the best solution because of compatibility, thanks for pointing that. Fortunately I found a matlab function which does exactly the same. So in this last commit, in readSpikeTrials_timelocked, I changed the function which does binary search, and I removed all the cfg.minbadtime.
Before removing the 'test' part of this script and making the same changes in readLFP and in readSpikeTrials_windowed, let me know if this solution suits you and if it works for you :)

Paul-Baudin committed:
Delete binarySearch.mexw64

Paul-Baudin committed:
fix bug in endtime trialinfo, and if artefact overlaps several trials
